### PR TITLE
add middleware to role base authen

### PR DESCRIPTION
**Pull request by chinttmor**
# Description
- add middleware to verify user role
# Evidence

[Screencast from 23-01-2024 16:04:27.webm](https://github.com/chinttmor/MMP-UI/assets/149034455/5175ebbe-61e1-47da-b00d-631062afc4d1)

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] My changes do not affect other functions
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/app/api/auth/[...nextauth]/options.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            },
+          },
+        );
+        let res = response.data;
+        if (res.statusCode === 201) {
+          // Any object returned will be saved in `user` property of the JWT
+          let user: User = {
+            ...res.data,
+          };
+          console.log('user', user);
```
Comment (mor-manhnd) [inline src/app/api/auth/[...nextauth]/options.ts:42]: remove it

File: src/app/auth/register/page.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+          phone: data.phone,
+        },
+        {
+          headers: {
+            'Content-Type': 'application/json',
+          },
+        },
+      );
+      return toast.success(`Data input is ok , sign in with your new account`, {
+        position: 'top-center',
```
Comment (mor-manhnd) [inline src/app/auth/register/page.tsx:37]: duplicate logic if toast

File: src/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+export async function middleware(request: NextRequest) {
+  const redirectUrl = request.nextUrl;
+  const hasSessionToken = request.cookies.has('next-auth.session-token');
+  const isSignInRoute = redirectUrl.pathname === '/auth/sign-in';
+  const protectedPaths = ['/admin'];
+  const matchesProtectedPath = protectedPaths.some((path) =>
+    redirectUrl.pathname.startsWith(path),
+  );
+  const token = await getToken({ req: request });
+  if (hasSessionToken) {
```
Comment (mor-manhnd) [inline src/middleware.ts:15]: const redirectUrl = new URL(request.url);
const isAdmin = hasSessionToken && matchesProtectedPath && token.zone[0] === Role.ADMIN;

if (isAdmin) {
  return NextResponse.next();
}

if (!hasSessionToken) {
  redirectUrl.pathname = '/auth/sign-in';
} else {
  redirectUrl.pathname = '/landing-page';
}

return NextResponse.redirect(redirectUrl);

**Review by mor-manhnd — APPROVED** [review]

**Comment by duc-lt-mor** [comment]
looks ok enough

**Pull request merged**